Sources/StreamChat/AudioPlayer/AudioPlayerObserving.swift

```diff
@@ -0,0 +1,136 @@
+    /// report changing time.
+    /// - Parameters:
+    ///   - player: The player from which we would like to receive updates
+    ///   - interval: The interval at which we would like to receive updates
+    ///   - queue: The queue on which the call to block will occur every time there is an update
+    ///   - block: The block to call every time there is an update
+    func addPeriodicTimeObserver(
+        _ player: AVPlayer,
+        forInterval interval: CMTime,
+        queue: DispatchQueue?,
+        using block: @escaping () -> Void
+    )
+
+    /// Registers and observer that will be called once the playback of an item stops
+    /// - Parameters:
+    ///   - queue: The queue on which the `stoppedPlaybackHandler` will be dispatched
+    ///   - block: The block to call once a player's item has stopped
+    func addStoppedPlaybackObserver(
+        queue: OperationQueue?,
+        using block: @escaping (AVPlayerItem) -> Void
+    )
+}
+
+public class StreamPlayerObserver: AudioPlayerObserving {
+    /// An observer token that is used to observe the ``AVPlayer.timeControlStatus`` of the player to receive
+    /// updates when the state changes between playing and paused.
+    private var timeControlStatusObserver: NSKeyValueObservation?
+
+    /// A token referencing the periodicTimer that is registered on the player and is used to provide
+    /// time-related metadata updates.
+    private var periodicTimeObservationToken: Any?
+
+    /// This block is called on deinit to remove any registered periodicTimeObserver
+    ///
+    /// According to the documentation we need to remove the periodicTimeObserver if we had
+    /// registered one.
+    /// https://developer.apple.com/documentation/avfoundation/avplayer/1385829-addperiodictimeobserverforinterv#return_value
+    private var periodicTimeObservationCancellationBlock: (() -> Void)?
+
+    /// An observer token that is used to reference the NotificationCenter registration, that is being used
+    /// to get notifications when the player's playback has been stopped.
+    private var stoppedPlaybackObservationToken: Any?
+
+    /// The notificationCenter on which the ``playbackFinishedObserver`` will be registered on
+    private let notificationCenter: NotificationCenter
+
+    // MARK: - Lifecycle
+
+    /// Creates a new instance of StreamPlayerObserver
+    /// - Parameter notificationCenter: The notificationCenter on which the stoppedPlaybackObserver
+    /// will be registered to listen for the `NSNotification.Name.AVPlayerItemDidPlayToEndTime`
+    /// notifications.
+    public init(
+        notificationCenter: NotificationCenter = .default
+    ) {
+        self.notificationCenter = notificationCenter
+    }
+
+    deinit {
+        timeControlStatusObserver?.invalidate()
+        periodicTimeObservationCancellationBlock?()
+        stoppedPlaybackObservationToken.map { notificationCenter.removeObserver($0) }
```

Comment:
No need to remove `periodicTimeObservationToken`?

Reply:
The `periodicTimeObservationToken` removal requires the player. In order to not hold a reference of it in the observer we are using the cancellationBlock which it does the following 
```
periodicTimeObservationCancellationBlock = { [weak player, periodicTimeObservationToken] in
            guard
                let player = player,
                let periodicTimeObservationToken = periodicTimeObservationToken
            else {
                return
            }
            player.removeTimeObserver(periodicTimeObservationToken)
        }
```

Reply:
Oh, missed this part! Thanks!